core/lib/web3_decl/src/client.rs

```diff
@@ -0,0 +1,456 @@
+}
+
+impl fmt::Display for RateLimitOrigin<'_> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Notification(name) => write!(formatter, "notification `{name}`"),
+            Self::Request(name) => write!(formatter, "request `{name}`"),
+            Self::BatchRequest(req) => {
+                let method_names: Vec<_> = req.iter().map(|(name, _)| name).collect();
+                write!(formatter, "batch request {method_names:?}")
+            }
+            Self::Subscription(name) => write!(formatter, "subscription `{name}`"),
+        }
+    }
+}
+
+/// JSON-RPC client for the main node with built-in middleware support.
+///
+/// The client should be used instead of `HttpClient` etc. A single instance of the client should be built
+/// and shared among all tasks run by the node in order to correctly rate-limit requests.
+#[derive(Debug, Clone)]
+pub struct L2Client {
+    inner: HttpClient,
+    rate_limit: SharedRateLimit,
+    component_name: &'static str,
+}
+
+impl L2Client {
+    pub fn builder() -> L2ClientBuilder {
+        L2ClientBuilder::default()
+    }
+
+    /// Sets the component operating this client. This is used in logging etc.
+    pub fn for_component(mut self, component_name: &'static str) -> Self {
+        self.component_name = component_name;
+        self
+    }
+
+    fn report_waiting(&self, origin: RateLimitOrigin, stats: &AcquireStats) {
+        if !stats.was_waiting {
+            return;
+        }
+
+        tracing::debug!(
+            component = self.component_name,
+            %origin,
+            "Request to {origin} by component `{}` was rate-limited using policy {} reqs/{:?}: {stats:?}",
+            self.component_name,
+            self.rate_limit.rate_limit,
+            self.rate_limit.rate_limit_window
+        );
+    }
+}
+
+#[async_trait]
+impl ClientT for L2Client {
+    async fn notification<Params>(&self, method: &str, params: Params) -> Result<(), Error>
+    where
+        Params: ToRpcParams + Send,
+    {
+        let stats = self.rate_limit.acquire(1).await;
+        self.report_waiting(RateLimitOrigin::Notification(method), &stats);
+        self.inner.notification(method, params).await
+    }
+
+    async fn request<R, Params>(&self, method: &str, params: Params) -> Result<R, Error>
+    where
+        R: DeserializeOwned,
+        Params: ToRpcParams + Send,
+    {
+        let stats = self.rate_limit.acquire(1).await;
+        self.report_waiting(RateLimitOrigin::Request(method), &stats);
+        self.inner.request(method, params).await
+    }
+
+    async fn batch_request<'a, R>(
+        &self,
+        batch: BatchRequestBuilder<'a>,
+    ) -> Result<BatchResponse<'a, R>, Error>
+    where
+        R: DeserializeOwned + fmt::Debug + 'a,
+    {
+        let stats = self.rate_limit.acquire(batch.iter().count()).await;
+        self.report_waiting(RateLimitOrigin::BatchRequest(&batch), &stats);
```

Comment:
I'd personally move those two lines (the same for other methods) into a method like throttle_if_needed (origin, requests_count)